src/drivers/imu/mpu9250/mag.cpp

```diff
@@ -205,19 +221,45 @@ MPU9250_mag::_measure(struct ak8963_regs data)
 
 	mag_report	mrb;
 	mrb.timestamp = hrt_absolute_time();
-	mrb.is_external = false;
+//	mrb.is_external = false;
+
+	// need a better check here. Using _parent->is_external() for mpu9250 also sets the
+	// internal magnetometers connected to the "external" spi bus as external, at least
+	// on Pixhawk 2.1. For now assuming the ICM20948 is only used on Here GPS, hence external.
+	if (_parent->_device_type == MPU_DEVICE_TYPE_ICM20948) {
+		mrb.is_external = _parent->is_external();
+
+	} else {
+		mrb.is_external = false;
+	}
 
 	/*
 	 * Align axes - note the accel & gryo are also re-aligned so this
 	 *              doesn't look obvious with the datasheet
 	 */
-	mrb.x_raw =  data.x;
-	mrb.y_raw = -data.y;
-	mrb.z_raw = -data.z;
+	float xraw_f, yraw_f, zraw_f;
+
+	if (_parent->_device_type == MPU_DEVICE_TYPE_ICM20948) {
+		/*
+		 * Keeping consistent with the accel and gyro axes of the ICM20948 here, just aligning the magnetometer to them.
+		 */
+		mrb.x_raw = data.y;
+		mrb.y_raw = data.x;
+		mrb.z_raw = -data.z;
+
+		xraw_f = data.y;
+		yraw_f = data.x;
+		zraw_f = -data.z;
+
+	} else {
+		mrb.x_raw =  data.x;
+		mrb.y_raw = -data.y;
+		mrb.z_raw = -data.z;
 
-	float xraw_f =  data.x;
-	float yraw_f = -data.y;
-	float zraw_f = -data.z;
+		xraw_f =  data.x;
+		yraw_f = -data.y;
+		zraw_f = -data.z;
+	}
 
 	/* apply user specified rotation */
 	rotate_3f(_parent->_rotation, xraw_f, yraw_f, zraw_f);
```

Comment:
@flochir I think we should add the rotation of the mag with respect to accel and gyro here. So that when we start the driver with -R 6 we don't need to correct the compass rotation on qgc (this should just be used if the gps module is rotated with respect to the front of the vehicle)